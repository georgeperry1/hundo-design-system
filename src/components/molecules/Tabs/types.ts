export const TabVariant = {
  Primary: 'primary',
  Secondary: 'secondary',
  Icon: 'icon',
};
