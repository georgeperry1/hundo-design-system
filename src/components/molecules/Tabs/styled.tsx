import React, { CSSProperties } from 'react';
import styled from 'styled-components';

import { TabVariant } from './types';

type TabBarProps = {
  hideDivider?: boolean;
  align?: CSSProperties['justifyContent'];
  variant?: TabVariant;
};

const PrimaryTabBar = styled.div<TabBarProps>`
  display: flex;
  align-items: center;
  justify-content: ${({ align }) => align};
  border-bottom: 1px solid;
  border-bottom-color: ${({ theme, hideDivider }) => {
    return hideDivider ? theme.colors.fullWhite : theme.colors.grey04;
  }};
`;

const SecondaryTabBar = styled.div<TabBarProps>`
  display: flex;
  align-items: center;
  justify-content: ${({ align }) => align};
  background: transparent;
`;

export const TabBar: React.FC<React.PropsWithChildren<TabBarProps>> = ({ hideDivider, align, variant, children }) => {
  if (variant === TabVariant.Secondary) {
    return <SecondaryTabBar align={align}>{children}</SecondaryTabBar>;
  }

  return (
    <PrimaryTabBar hideDivider={hideDivider} align={align}>
      {children}
    </PrimaryTabBar>
  );
};
