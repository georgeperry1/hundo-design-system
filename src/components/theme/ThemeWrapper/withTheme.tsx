import React from 'react';
import { ThemeProvider } from 'styled-components';
import { lightTheme } from '../themes';

export function withTheme<P>(WrappedComponent: React.ComponentType<P>): React.FC<P> {
  const ComponentWithTheme = (props: P) => {
    return (
      <ThemeProvider theme={lightTheme}>
        <WrappedComponent {...props} />
      </ThemeProvider>
    );
  };
  return ComponentWithTheme;
}
export default withTheme;
