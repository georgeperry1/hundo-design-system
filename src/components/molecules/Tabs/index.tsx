import React, { CSSProperties, useState, useEffect } from 'react';

import { TabBar } from './styled';
import { TabItem } from '../../atoms';
import { TabVariant } from './types';

export type TabOption = {
  value: string;
  label: string;
  icon?: string;
  disabled?: boolean;
  variant?: string;
};

export interface TabsProps {
  initialValue?: string;
  value?: string;
  tabOptions: TabOption[];
  hideDivider?: boolean;
  hideBorder?: boolean;
  onChange?: (val: string) => void;
  align?: CSSProperties['justifyContent'];
  variant?: string;
}

const Tabs: React.FC<React.PropsWithChildren<TabsProps>> = ({
  initialValue,
  value,
  hideDivider,
  hideBorder,
  onChange,
  tabOptions,
  align,
  variant,
}) => {
  const [currentTab, setCurrentTab] = useState<string | undefined>(initialValue);

  useEffect(() => {
    if (typeof value === 'undefined') return;
    setCurrentTab(value);
  }, [value]);

  const onTabClick = (tab: string) => {
    setCurrentTab(tab);
    onChange && onChange(tab);
  };

  return (
    <TabBar hideDivider={hideDivider} align={align} variant={variant}>
      {tabOptions &&
        tabOptions.map(({ value, ...rest }) => (
          <TabItem
            id={`id-${value}`}
            onTabClick={() => onTabClick(value)}
            value={value}
            key={value}
            active={currentTab === value}
            hideBorder={hideBorder}
            variant={variant}
            totalTabs={tabOptions.length}
            {...rest}
          />
        ))}
    </TabBar>
  );
};

Tabs.defaultProps = {
  hideDivider: false,
  hideBorder: false,
  align: 'flex-start',
  variant: TabVariant.Primary,
};

export default Tabs;
