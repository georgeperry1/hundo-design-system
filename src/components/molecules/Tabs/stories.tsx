import { Story, Meta } from '@storybook/react';

import Tabs, { TabsProps } from '.';

export default {
  title: 'Components/Molecules/Tabs',
  component: Tabs,
} as Meta;

const Template: Story<TabsProps> = (args) => <Tabs {...args} />;

export const Default = Template.bind({});

Default.args = {
  tabOptions: [
    { value: 'Tab 1', label: 'Tab 1' },
    { value: 'Tab 2', label: 'Tab 2' },
    { value: 'Tab 3', label: 'Tab 3' },
  ],
};

export const InitialValue = Template.bind({});

InitialValue.args = {
  tabOptions: [
    { value: 'Tab 1', label: 'Tab 1' },
    { value: 'Tab 2', label: 'Tab 2' },
    { value: 'Tab 3', label: 'Tab 3' },
  ],
  initialValue: 'Tab 2',
};

export const HideDivider = Template.bind({});

HideDivider.args = {
  tabOptions: [
    { value: 'Tab 1', label: 'Tab 1' },
    { value: 'Tab 2', label: 'Tab 2' },
    { value: 'Tab 3', label: 'Tab 3' },
  ],
  initialValue: 'Tab 2',
  hideDivider: true,
};

export const Align = Template.bind({});

Align.args = {
  tabOptions: [
    { value: 'Tab 1', label: 'Tab 1' },
    { value: 'Tab 2', label: 'Tab 2' },
    { value: 'Tab 3', label: 'Tab 3' },
  ],
  initialValue: 'Tab 2',
  align: 'flex-end',
};
