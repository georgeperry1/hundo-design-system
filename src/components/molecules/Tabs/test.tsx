import { render, fireEvent, screen } from '@testing-library/react';

import Tabs from '.';
import ThemeWrapper from '../../../utils/test-theme-wrapper';

const mockTabOptions = [
  { value: 'Tab 1', label: 'Tab 1' },
  { value: 'Tab 2', label: 'Tab 2' },
  { value: 'Tab 3', label: 'Tab 3' },
];

describe('Spacer', () => {
  it('should render', () => {
    const { baseElement } = render(
      <ThemeWrapper TestedComponent={Tabs} componentProps={{ tabOptions: mockTabOptions }} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should show an active tab when an initial value is passed', () => {
    const { baseElement } = render(
      <ThemeWrapper TestedComponent={Tabs} componentProps={{ tabOptions: mockTabOptions, initialValue: 'Tab 1' }} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should hide the divider', () => {
    const { baseElement } = render(
      <ThemeWrapper TestedComponent={Tabs} componentProps={{ tabOptions: mockTabOptions, hideDivider: true }} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should change the alignments of the tabs', () => {
    const { baseElement } = render(
      <ThemeWrapper TestedComponent={Tabs} componentProps={{ tabOptions: mockTabOptions, align: 'flex-end' }} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should send an event when changed', () => {
    const handleChange = jest.fn();

    render(
      <ThemeWrapper TestedComponent={Tabs} componentProps={{ tabOptions: mockTabOptions, onChange: handleChange }} />,
    );

    fireEvent.click(screen.getByText('Tab 1'));

    expect(handleChange).toHaveBeenCalledWith('Tab 1');
  });
});
