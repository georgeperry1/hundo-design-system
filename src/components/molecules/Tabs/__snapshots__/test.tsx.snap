// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer should change the alignments of the tabs 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c1 {
  background: #ffffff;
  border-bottom: 1px solid;
  border-bottom-color: #BDBDBD;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  opacity: 100%;
}

.c3 {
  min-height: 35px;
  padding: 0 12px;
  border-bottom: 4px solid;
  border-bottom-color: #ffffff;
}

.c3:hover {
  border-bottom-color: #000000;
}

.c4 {
  font-family: Inter;
  font-size: 14px;
  font-weight: 500;
  line-height: 24px;
  margin: 10px 0;
  color: #000000;
}

<body>
  <div>
    <div
      class="c0 c1"
    >
      <div
        class="c2 c3"
        value="Tab 1"
      >
        <p
          class="c4"
        >
          Tab 1
        </p>
      </div>
      <div
        class="c2 c3"
        value="Tab 2"
      >
        <p
          class="c4"
        >
          Tab 2
        </p>
      </div>
      <div
        class="c2 c3"
        value="Tab 3"
      >
        <p
          class="c4"
        >
          Tab 3
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Spacer should hide the divider 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c1 {
  background: #ffffff;
  border-bottom: 1px solid;
  border-bottom-color: #ffffff;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  opacity: 100%;
}

.c3 {
  min-height: 35px;
  padding: 0 12px;
  border-bottom: 4px solid;
  border-bottom-color: #ffffff;
}

.c3:hover {
  border-bottom-color: #000000;
}

.c4 {
  font-family: Inter;
  font-size: 14px;
  font-weight: 500;
  line-height: 24px;
  margin: 10px 0;
  color: #000000;
}

<body>
  <div>
    <div
      class="c0 c1"
    >
      <div
        class="c2 c3"
        value="Tab 1"
      >
        <p
          class="c4"
        >
          Tab 1
        </p>
      </div>
      <div
        class="c2 c3"
        value="Tab 2"
      >
        <p
          class="c4"
        >
          Tab 2
        </p>
      </div>
      <div
        class="c2 c3"
        value="Tab 3"
      >
        <p
          class="c4"
        >
          Tab 3
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Spacer should render 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c1 {
  background: #ffffff;
  border-bottom: 1px solid;
  border-bottom-color: #BDBDBD;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  opacity: 100%;
}

.c3 {
  min-height: 35px;
  padding: 0 12px;
  border-bottom: 4px solid;
  border-bottom-color: #ffffff;
}

.c3:hover {
  border-bottom-color: #000000;
}

.c4 {
  font-family: Inter;
  font-size: 14px;
  font-weight: 500;
  line-height: 24px;
  margin: 10px 0;
  color: #000000;
}

<body>
  <div>
    <div
      class="c0 c1"
    >
      <div
        class="c2 c3"
        value="Tab 1"
      >
        <p
          class="c4"
        >
          Tab 1
        </p>
      </div>
      <div
        class="c2 c3"
        value="Tab 2"
      >
        <p
          class="c4"
        >
          Tab 2
        </p>
      </div>
      <div
        class="c2 c3"
        value="Tab 3"
      >
        <p
          class="c4"
        >
          Tab 3
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Spacer should show an active tab when an initial value is passed 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c1 {
  background: #ffffff;
  border-bottom: 1px solid;
  border-bottom-color: #BDBDBD;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  opacity: 100%;
}

.c3 {
  min-height: 35px;
  padding: 0 12px;
  border-bottom: 4px solid;
  border-bottom-color: #000000;
}

.c3:hover {
  border-bottom-color: #000000;
}

.c5 {
  min-height: 35px;
  padding: 0 12px;
  border-bottom: 4px solid;
  border-bottom-color: #ffffff;
}

.c5:hover {
  border-bottom-color: #000000;
}

.c4 {
  font-family: Inter;
  font-size: 14px;
  font-weight: 500;
  line-height: 24px;
  margin: 10px 0;
  color: #000000;
}

<body>
  <div>
    <div
      class="c0 c1"
    >
      <div
        class="c2 c3"
        value="Tab 1"
      >
        <p
          class="c4"
        >
          Tab 1
        </p>
      </div>
      <div
        class="c2 c5"
        value="Tab 2"
      >
        <p
          class="c4"
        >
          Tab 2
        </p>
      </div>
      <div
        class="c2 c5"
        value="Tab 3"
      >
        <p
          class="c4"
        >
          Tab 3
        </p>
      </div>
    </div>
  </div>
</body>
`;
